src/Components/Server/src/Circuits/RemoteJSDataStream.cs

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    internal sealed class RemoteJSDataStream : Stream
+    {
+        private readonly RemoteJSRuntime _runtime;
+        private readonly long _streamId;
+        private readonly long _totalLength;
+        private readonly CancellationToken _streamCancellationToken;
+        private readonly Stream _pipeReaderStream;
+        private readonly Pipe _pipe;
+        private long _bytesRead;
+
+        public static async Task<bool> ReceiveData(RemoteJSRuntime runtime, long streamId, byte[] chunk, string error)
+        {
+            if (!runtime.RemoteJSDataStreamInstances.TryGetValue(streamId, out var instance))
+            {
+                // There is no data stream with the given identifier. It may have already been disposed.
+                // We notify JS that the stream has been cancelled/disposed.
+                return false;
+            }
+
+            return await instance.ReceiveData(chunk, error);
+        }
+
+        public static async ValueTask<RemoteJSDataStream> CreateRemoteJSDataStreamAsync(
+            RemoteJSRuntime runtime,
+            IJSDataReference jsDataReference,
+            long totalLength,
+            long maxBufferSize,
+            long maximumIncomingBytes,
+            CancellationToken cancellationToken = default)
+        {
+            var streamId = runtime.RemoteJSDataStreamNextInstanceId++;
```

Comment:
Can this be called in parallel?

Reply:
Nope, we're operating under the single-threaded assumption throughout.

Reply:
This code runs in the Blazor server sync context which ensures only one thread runs at a time.